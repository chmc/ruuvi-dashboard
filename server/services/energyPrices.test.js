const energyPricesFromApi = require('./energyPricesFromApi')
const storage = require('../storage')
const energyPricesService = require('./energyPrices')

describe('energyPricesService', () => {
  describe('getEnergyPrices()', () => {
    beforeAll(() => {
      jest.useFakeTimers()
      jest.resetModules()
    })

    afterAll(() => {
      jest.useRealTimers()
    })

    it('should return object with energy prices for today', async () => {
      // Arrange
      jest.setSystemTime(new Date('2023-10-27T12:11:00'))
      const apiResponseJson = `
      [{
        "Rank": 1,
        "DateTime": "2023-10-27T00:00:00+03:00",
        "PriceNoTax": 0.0342,
        "PriceWithTax": 0.0424
      },
      {
        "Rank": 8,
        "DateTime": "2023-10-27T01:00:00+03:00",
        "PriceNoTax": 0.0509,
        "PriceWithTax": 0.0631
      },
      {
        "Rank": 7,
        "DateTime": "2023-10-27T02:00:00+03:00",
        "PriceNoTax": 0.0497,
        "PriceWithTax": 0.0617
      },
      {
        "Rank": 4,
        "DateTime": "2023-10-27T03:00:00+03:00",
        "PriceNoTax": 0.0486,
        "PriceWithTax": 0.0603
      }]`
      jest.spyOn(storage, 'loadOrDefault').mockResolvedValue(undefined)
      jest.spyOn(storage, 'save').mockImplementation(jest.fn)
      const getEnergyPricesFromApiSpy = jest.spyOn(
        energyPricesFromApi,
        'getEnergyPricesFromApi'
      )
      getEnergyPricesFromApiSpy.mockResolvedValue(apiResponseJson)

      // Act
      const result = await energyPricesService.getEnergyPrices(undefined)

      // Assert
      expect(getEnergyPricesFromApiSpy).toBeCalledTimes(1)
      expect(result).toMatchSnapshot()
    })

    it('should return current energy prices and not get new', async () => {
      // Arrange
      jest.setSystemTime(new Date('2023-10-24T12:11:00'))

      /** @type {EnergyPrices} */
      const energyPrices = {
        updatedAt: new Date('2023-10-24T12:00:00'),
        todayEnergyPrices: {
          updatedAt: new Date('2023-10-24T12:00:00'),
          pricesForDate: '2023-10-24',
          data: [
            {
              date: new Date('2023-10-24T12:00:00'),
              price: 0.05,
              hour: 12,
            },
          ],
        },
      }
      jest.spyOn(storage, 'loadOrDefault').mockResolvedValue(undefined)
      jest.spyOn(storage, 'save').mockImplementation(jest.fn)

      // Act
      const result = await energyPricesService.getEnergyPrices(energyPrices)

      // Assert
      expect(result).toMatchSnapshot()
    })
  })
})
