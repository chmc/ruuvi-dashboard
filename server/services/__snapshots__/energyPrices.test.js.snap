// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`energyPricesService getEnergyPrices() should return current energy prices and not get new 1`] = `
Object {
  "todayEnergyPrices": Object {
    "data": Array [
      Object {
        "date": 2023-10-24T09:00:00.000Z,
        "hour": 12,
        "price": 0.05,
      },
    ],
    "pricesForDate": "2023-10-24",
    "updatedAt": 2023-10-24T09:00:00.000Z,
  },
  "tomorrowEnergyPrices": undefined,
  "updatedAt": 2023-10-24T09:00:00.000Z,
}
`;

exports[`energyPricesService getEnergyPrices() should return object with energy prices for today 1`] = `
Object {
  "todayEnergyPrices": Object {
    "data": Array [
      Object {
        "date": 2023-10-26T21:00:00.000Z,
        "hour": 0,
        "price": 4.24,
      },
      Object {
        "date": 2023-10-26T22:00:00.000Z,
        "hour": 1,
        "price": 6.31,
      },
      Object {
        "date": 2023-10-26T23:00:00.000Z,
        "hour": 2,
        "price": 6.17,
      },
      Object {
        "date": 2023-10-27T00:00:00.000Z,
        "hour": 3,
        "price": 6.03,
      },
    ],
    "pricesForDate": "2023-10-27",
    "updatedAt": 2023-10-27T09:11:00.000Z,
  },
  "tomorrowEnergyPrices": undefined,
  "updatedAt": 2023-10-27T09:11:00.000Z,
}
`;
